### ISSUE TITLE
MW1.31: Add/Edit category box is empty

### ISSUE THREAD

**Issue by wambaloo** [Mediawiki-wysiwyg/WYSIWYG-CKeditor#116]
Hi !

I faced a problem and I don't know if it's a bug or a misconfiguration.

MediaWiki | 1.31.0
-- | --
PHP | 7.2
-- | --
WikiEditor | 0.5.1

Here is my LocalSettings.php file : 
``` php
#Default user options:
$wgDefaultUserOptions['riched_disable']               = false;
$wgDefaultUserOptions['riched_start_disabled']        = false;
$wgDefaultUserOptions['riched_use_toggle']            = true; 
$wgDefaultUserOptions['riched_use_popup']             = false;
$wgDefaultUserOptions['riched_toggle_remember_state'] = true;
$wgDefaultUserOptions['riched_link_paste_text']       = true;

//MW<=1.24 or versions of WYSIWYG <= "1.5.6_0 [B551++01.07.2016]"
require_once "$IP/extensions/WYSIWYG/WYSIWYG.php"; 

//MW>=1.25 and versions of WYSIWYG >= "1.5.6_0 [B551+02.07.2016]"
wfLoadExtension( 'WYSIWYG' );

//MW>=1.25 and versions of WYSIWYG >= "1.5.6_0 [B551+02.07.2016]" has dependency
//to module of WikiEditor so it must be enabled too (or otherwise file
//extension.json has to be edited manually to remove dependency)
wfLoadExtension( 'WikiEditor' );
```

But when I click on the category button, the box is empty and search do nothing
![image](https://user-images.githubusercontent.com/15379206/45037747-f561a900-b05f-11e8-9065-b47ea65f402a.png)

How can I solve it please ?

Thanks in advance.

**Comment by riparap** [comment]
If in category dialog you type something in search filter box, it should populate the list of categories with category names matching your value of filter. 

For some reason db query made by wysiwyg for categories fails in MW 1.31. At this moment I am not sure what is causing this and how it can be fixed.

**Comment by riparap** [comment]
This is somehow related to collate settings of database.

LocalSettings.php:
`$wgDBTableOptions = "ENGINE=InnoDB, DEFAULT CHARSET=binary";`

Database:
``sql:> SHOW VARIABLES LIKE "%char%";`` 
 
`` +--------------------------+----------------------------+`` 
`` | Variable_name            | Value                      |`` 
`` +--------------------------+----------------------------+`` 
`` | character_set_client     | utf8                       |`` 
`` | character_set_connection | utf8                       |`` 
`` | character_set_database   | latin1                     |`` 
`` | character_set_filesystem | binary                     |`` 
`` | character_set_results    | utf8                       |`` 
`` | character_set_server     | latin1                     |`` 
`` | character_set_system     | utf8                       |`` 
`` | character_sets_dir       | /opt/lampp/share/charsets/ |`` 
`` +--------------------------+----------------------------+`` 

Sql query for categories fails because of the "COLLATE utf8_unicode_ci" definition in query:
`ERROR 1253 (42000): COLLATION 'utf8_unicode_ci' is not valid for CHARACTER SET 'binary'`

Workaround: remove definition `COLLATE utf8_unicode_ci` from query.

In file
``   mediawiki\extensions\WYSIWYG\CKeditorSajax.body.php`` 
in function 
``   wfSajaxSearchCategoryCKeditor`` 
where variable $m_sql is set for sql query, in last line for quere, change
``  "GROUP BY tmpSelectCat1.cl_to COLLATE utf8_unicode_ci"; `` 
to  
``   "GROUP BY tmpSelectCat1.cl_to";  // COLLATE utf8_unicode_ci"; `` 

I am not sure how this should be fixed properly. It is possible to set collate settings also in `my.cnf` file on server side so maybe `COLLATE` definition in query is in wrong place and should be removed permanently. Any ideas about this ?

Workaround has been included in Version 1.5.6_0 [+18.09.2018].

Also workaround for situation where nested categories created infinite loop when building list of categories have been included in this version (by HellToupee1).

**Comment by hannes1985** [comment]
Hi everybody,
I have a little problem. 
When I click in the editor to "source" (german: Quellcode), then there is always a "empty-space" or "line-break" character in first and last line. So my formation in the edited wiki-page does not look as before. So i always have to delete the empty-line in first and last row. It also appears, when i switch on or off the source-code editor.

**See the following screenshot:**
![grafik](https://user-images.githubusercontent.com/43953824/46602364-3f6efc00-caf0-11e8-8479-8578f69dbd28.png)

**Systeminfo:**

MediaWiki | 1.30.0
-- | --
PHP | 5.6.24
-- | --
WikiEditor | 0.5.1
